graphix_zx/graphstate.py

```diff
@@ -695,45 +695,75 @@ def compose_sequentially(  # noqa: C901
     if set(graph1.output_node_indices.values()) != set(graph2.input_node_indices.values()):
         msg = "Logical qubit indices of output nodes in graph1 must match input nodes in graph2."
         raise ValueError(msg)
-    node_map1: dict[int, int] = {}
-    node_map2: dict[int, int] = {}
+
     composed_graph = GraphState()
 
-    for node in graph1.physical_nodes - graph1.output_node_indices.keys():
-        node_index = composed_graph.add_physical_node()
-        meas_basis = graph1.meas_bases.get(node, None)
-        if meas_basis is not None:
-            composed_graph.assign_meas_basis(node_index, meas_basis)
-        node_map1[node] = node_index
+    node_map1 = _copy_nodes(
+        src=graph1,
+        dst=composed_graph,
+        exclude_nodes=set(graph1.output_node_indices.keys()),
```

Comment:
You can directly pass the dict key as a AbstractSet.

```suggestion
        exclude_nodes=graph1.output_node_indices,
```

Reply:
Sorry I misunderstood. `.keys()`  is necessary but `set` collection is redundant